### PR TITLE
Tracking control-flow linearity for effect handlers

### PR DESCRIPTION
**Pull request by thwfhk**
This PR will track the control-flow linearity for effect handlers, which will resolve issue #544.

It is very incomplete and full of hacks. More explanation to be added. Currently I just want to create a PR to be able to sync with the main branch.

### CODE REVIEW

File: core/typeSugar.ml
Comment (dhil) [inline core/typeSugar.ml:2946]:
```suggestion
            (* make types of parameters unlimited if they are not used exactly once *)
```

File: core/typeSugar.ml
Comment (dhil) [inline core/typeSugar.ml:4012]: Dead code?

File: core/typeSugar.ml
Comment (dhil) [inline core/typeSugar.ml:4317]:
```suggestion
```

File: core/typeSugar.ml
Comment (dhil) [inline core/typeSugar.ml:5218]: Dead code?

File: core/typeSugar.ml
Comment (dhil) [inline core/typeSugar.ml:5424]: Dead code?

File: core/types.ml
Comment (dhil) [inline core/types.ml:3136]: Yes, this should probably be updated. This logic here is used for pretty-printing types. It reintroduces some of the sugar that was elaborated by earlier passes.

File: core/types.mli
Comment (dhil) [inline core/types.mli:147]:
```suggestion
```

File: core/desugarSessionExceptions.ml
Comment (dhil) [reply core/desugarSessionExceptions.ml:90]: Though, I see that you have changed the signature of `SessionFail` below... I would be inclined not to change any of the session exceptions stuff.

**Review by dhil — COMMENTED** [review]
LGTM. There are some extraneous comments that should be removed before merge.

**Comment by dhil** [comment]
Another thing is that you should hide this feature behind a new flag, e.g. `track_control_flow_linearity`. If that flag is not enabled, then you should emit an error on uses of any of the features included in this PR.

File: core/desugarEffects.ml
Comment (thwfhk) [reply core/desugarEffects.ml:368]: I see. Then treating it as an unlimited operation is fine.

File: core/parser.mly
Comment (thwfhk) [reply core/parser.mly:914]: Yes. Removed.

File: core/sugartypes.ml
Comment (thwfhk) [reply core/sugartypes.ml:201]: Sounds good. I'll do this later.

File: core/typeSugar.ml
Comment (thwfhk) [reply core/typeSugar.ml:1903]: Removed.

File: core/typeSugar.ml
Comment (thwfhk) [reply core/typeSugar.ml:2774]: Removed

File: core/typeSugar.ml
Comment (thwfhk) [reply core/typeSugar.ml:2479]: Then why all other cases return an operation type, but this case returns a tuple type? Is this followed by some desugaring? The problem is I need to find somewhere to put the `is_lincase`.

File: core/typeSugar.ml
Comment (thwfhk) [reply core/typeSugar.ml:4000]: removed

File: core/typeSugar.ml
Comment (thwfhk) [reply core/typeSugar.ml:5206]: removed

File: core/types.ml
Comment (thwfhk) [reply core/types.ml:3136]: One possible easy change here is to use `tp` instead of `c` for linear nullary operations. For example, `Hi : Bool` is an unlimited nullary operation, and `Hi : () =@ Bool` is a linear nullary operation.

File: core/sugartoir.ml
Comment (thwfhk) [reply core/sugartoir.ml:1005]: The `type_check` in `typeSugar.ml` desugars (by simply discarding the keyword) the `Unlet` and `Linlet`. This may not be the best place to do that.

File: core/desugarSessionExceptions.ml
Comment (thwfhk) [reply core/desugarSessionExceptions.ml:90]: I have to change the `SessionFail` operation to be linear to make the functions in `prelude.links` well-typed. Otherwise we would get a type error because we use the unlimited operation `SessionFail` with linear channels. I also need to change some the signatures of some functions in `lib.ml`, because by default `e::Row` is `e::Row(Unl,Any)`, which can only be unified with linear operations.

This is also a challenge for hiding this feature behind a new flag; it is not entirely compatible with the current Links syntax.

File: core/types.ml
Comment (dhil) [reply core/types.ml:3136]: Yes, maybe we should get rid of the nullary sugar altogether... But as a first step lets be explicit when the operation interpretation is linear.

File: core/desugarSessionExceptions.ml
Comment (dhil) [reply core/desugarSessionExceptions.ml:90]: I see... The settings logic does allow you specify transitive dependencies, so if session exceptions are toggled, then you can forcibly toggle linear continuations too.

Comment (thwfhk) [reply core/desugarSessionExceptions.ml:90]: Actually it is fine to not toggle control_flow_linearity, but we still need to make sure that the existing library code is compatible with it. FOr example, using unlimited row variable `e::Row(Any, Any)` by default and use the linear operation signature `=@` for `SessionFail`.

**Comment by thwfhk** [comment]
There is one remaining problem (or two):

* The linearity kinds of row variables reuse `Unl` and `Any`, where `e::Unl` means the row variable `e` can only be unified with linear operations, and `e::Any` means the row variable `e` can be unified with any operations.
* As a result, we need to change the the type signatures of library functions. For example, the type of plus should be `"(Int, Int) { |_::Any}-> Int"` instead of `(Int, Int) -> Int`, because the latter is less general as it can only be used with linear operations. We can also change the parsing and printing to use `->` to represent `{ |_::Any}->` by default.

File: core/typeSugar.ml
Comment (dhil) [reply core/typeSugar.ml:2479]: I suggest experimenting with this in a separate patch. 

Regarding `TypeUtils.arg_types`: I believe this is because we always represent the domain of function types as a record (recall functions are `n`-ary in Links). More specifically, it is a tuple, as the labels are `1` through `n`. We represent tuples as records.

**Comment by SimonJF** [comment]
@vcgalpin and @SimonJF are going to try this out on our examples

**Review by vcgalpin — COMMENTED** [review]
I have tested this branch with three different Links applications:
- Guide to Pharmacology DB (@SimonJF)
- https://github.com/vcgalpin/dcc-xml-temporal/
- https://github.com/vcgalpin/links-covid-curation

They typechecked and showed no errors under testing.

I can do more testing if there anything in particular that might cause a problem. (Note that I have not reviewed the code.)

**Review by dhil — APPROVED** [review]
LGTM. It is a priority to merge this soon.

**Review by dhil — APPROVED** [review]

**Review by dhil — APPROVED** [review]

**Comment by dhil** [comment]
@thwfhk there are a couple of failures now. I will merge this PR once the regressions have been fixed.

**Comment by thwfhk** [comment]
> @thwfhk there are a couple of failures now. I will merge this PR once the regressions have been fixed.

Ah yes. That's because I forgot to add the files of some new tests. They should be fixed now.

**Comment by dhil** [comment]
I'll merge after the CI has run successfully to completion.

**Comment by dhil** [comment]
> So... does this close #544?

No as per the meeting yesterday. We keep the issue open for a potentially fun gimmick. We will close it during POPL.

**Pull request merged**
